### ISSUE TITLE
[Bug] Players are randomly assigned EP and GP values

### ISSUE THREAD

**Issue by mykkenny** [Alumian/CEPGP-Retail#52]
**The Error**
No error

**Describe the bug**
I am unsure at what point this occurred, possibly during Onyxia loot distribution, though it seemed more players had issues than loot was given. Essentially around a dozen or so players had been assigned EP they did not earn, or GP they had not spent. After Onyxia I cleaned it up by amending the values directly in the officer notes, and kept a close eye when distributing loot in Molten Core, but the bug did not occur again. 

At the start of the night I did use the Record EPGP feature, and also the Export to CSV feature. Not sure if related. The addon was initially set to give 5EP for killing Onyxia, which it did to all but 2-3 of the raid.

**To Reproduce**
Unknown.

**Debug Information**
<details><summary>Debug Info</summary><br />Version: 1.12.0<br /><br />GP Modifier: 10<br /><br />Base GP: 1<br /><br />Standby EP: True<br /><br />Standby Offline: True<br /><br />Standby Percent: 100<br /><br />Standby Accept Whispers: True<br /><br />Standby EP by Rank: False<br /><br />Standby EP Manual Delegation: True<br /><br />GUI for Loot: False<br /><br />Loot Response Keyphrase: !need<br /><br />Standby EP Whisper Keyphrase: standby<br /><br /><details><summary>Auto EP</summary>High Priestess Jeklik: True<br />Gothik the Harvester: True<br />Buru the Gorger: True<br />Wushoolay: True<br />Lord Kri: True<br />Emperor Vek'nilash: True<br />Ossirian the Unscarred: True<br />Golemagg the Incinerator: False<br />Grand Widow Faerlina: True<br />High Priestess Mar'li: True<br />Majordomo Executus: False<br />High Priest Venoxis: True<br />Gehennas: False<br />Taerar: True<br />Ysondre: True<br />Thane Korth'azz: True<br />High Priest Thekal: True<br />Noth the Plaguebringer: True<br />Lethon: True<br />Moam: True<br />The Prophet Skeram: True<br />Flamegor: True<br />C'Thun: True<br />General Rajaxx: True<br />Princess Huhuran: True<br />Thaddius: True<br />Loatheb: True<br />Lady Blaumeux: True<br />Highlord Mograine: True<br />Heigan the Unclean: True<br />Azuregos: True<br />Doom Lord Kazzak: True<br />Teremus the Devourer: True<br />Jin'do the Hexxer: True<br />Baron Geddon: False<br />Renataki: True<br />Sulfuron Harbinger: False<br />Gri'lek: True<br />Bloodlord Mandokir: True<br />Patchwerk: True<br />Vaelastrasz the Corrupt: True<br />Gluth: True<br />Sapphiron: True<br />Emeriss: True<br />Onyxia: False<br />Princess Yauj: True<br />Kurinnaxx: True<br />Shazzrah: False<br />Ebonroc: True<br />Battleguard Sartura: True<br />Hakkar: True<br />Kel'Thuzad: True<br />Nefarian: True<br />Fankriss the Unyielding: True<br />Lucifron: False<br />Anub'Rekhan: True<br />Hazza'rah: True<br />Viscidus: True<br />Broodlord Lashlayer: True<br />Ayamiss the Hunter: True<br />Ragnaros: False<br />Maexxna: True<br />Ouro: True<br />Vem: True<br />Grobbulus: True<br />Instructor Razuvious: True<br />Razorgore the Untamed: True<br />Magmadar: False<br />Sir Zeliek: True<br />Gahz'ranka: True<br />Emperor Vek'lor: True<br />High Priestess Arlokk: True<br />Firemaw: True<br />Chromaggus: True<br />Garr: False<br /></details><br /><details><summary>EP Values</summary>High Priestess Jeklik: 2<br />Gothik the Harvester: 12<br />Buru the Gorger: 3<br />Wushoolay: 2<br />Lord Kri: 10<br />Emperor Vek'nilash: 10<br />Ossirian the Unscarred: 4<br />Golemagg the Incinerator: 5<br />Grand Widow Faerlina: 12<br />High Priestess Mar'li: 2<br />Majordomo Executus: 5<br />High Priest Venoxis: 2<br />Gehennas: 5<br />Taerar: 7<br />Ysondre: 7<br />Thane Korth'azz: 15<br />High Priest Thekal: 2<br />Noth the Plaguebringer: 12<br />Lethon: 7<br />Moam: 3<br />The Prophet Skeram: 10<br />Flamegor: 7<br />C'Thun: 12<br />General Rajaxx: 3<br />Princess Huhuran: 10<br />Thaddius: 15<br />Loatheb: 15<br />Lady Blaumeux: 15<br />Highlord Mograine: 15<br />Heigan the Unclean: 12<br />Azuregos: 7<br />Doom Lord Kazzak: 7<br />Teremus the Devourer: 7<br />Jin'do the Hexxer: 2<br />Baron Geddon: 5<br />Renataki: 2<br />Sulfuron Harbinger: 5<br />Gri'lek: 2<br />Bloodlord Mandokir: 2<br />Patchwerk: 12<br />Vaelastrasz the Corrupt: 7<br />Gluth: 12<br />Sapphiron: 15<br />Emeriss: 7<br />Onyxia: 5<br />Princess Yauj: 10<br />Kurinnaxx: 3<br />Shazzrah: 5<br />Ebonroc: 7<br />Battleguard Sartura: 10<br />Hakkar: 3<br />Kel'Thuzad: 15<br />Nefarian: 10<br />Fankriss the Unyielding: 10<br />Lucifron: 5<br />Anub'Rekhan: 12<br />Hazza'rah: 2<br />Viscidus: 10<br />Broodlord Lashlayer: 7<br />Ayamiss the Hunter: 3<br />Ragnaros: 7<br />Maexxna: 15<br />Ouro: 10<br />Vem: 10<br />Grobbulus: 12<br />Instructor Razuvious: 12<br />Razorgore the Untamed: 7<br />Magmadar: 5<br />Sir Zeliek: 15<br />Gahz'ranka: 2<br />Emperor Vek'lor: 10<br />High Priestess Arlokk: 2<br />Firemaw: 7<br />Chromaggus: 7<br />Garr: 5<br /></details><br /><details><summary>Standby Guild Ranks</summary>Guild Master: False<br />Officer: False<br />Officer Alt: False<br />Raider: False<br />Member: False<br />Alt: False<br />Social Rank: False<br />Trial: False<br />Noob: False<br /></details><br /><details><summary>Slot Weights</summary>2HWEAPON: 2<br />NECK: 0.5<br />HEAD: 1<br />WEAPON: 1.5<br />WRIST: 0.5<br />WEAPONMAINHAND: 1.5<br />SHIELD: 0.5<br />HOLDABLE: 0.5<br />RANGED: 0.5<br />THROWN: 0.5<br />FEET: 0.75<br />RELIC: 0.5<br />RANGEDRIGHT: 0.5<br />WEAPONOFFHAND: 0.5<br />TRINKET: 0.75<br />FINGER: 0.5<br />CLOAK: 0.5<br />CHEST: 1<br />ROBE: 1<br />HAND: 0.75<br />WAIST: 0.75<br />EXCEPTION: 1<br />LEGS: 1<br />SHOULDER: 0.75<br /></details><br /><details><summary>Override List</summary></details><br /><details><summary>Addons List</summary><br />Atlas<br />Atlas_ClassicWoW<br />AtlasLootClassic<br />Auctionator<br />BetterVendorPrice<br />CEPGP<br />Clique<br />GigaAutoResourcesTracker<br />ItemTooltipProfessionIcons<br />Kui_Media<br />Kui_Nameplates<br />Kui_Nameplates_Core<br />Leatrix_Maps<br />Leatrix_Plus<br />Peggle<br />QuestLogEx<br />RealMobHealth<br />RecipeRadarClassic<br />Recount<br />TriviaBot<br />WhatsTraining<br />CharacterStatsClassic<br />ProjectAzilroka<br />AddOnSkins<br />ElvUI_BenikUI_Classic<br />ElvUI<br />DBM-Core<br />DBM-DefaultSkin<br />DBM-StatusBarTimers<br />Questie<br /></details></details>

**Comment by Alumian** [comment]
Hi Mykkenny,
Thanks for submitting this report. It sounds like the raid roster and guild rosters are desynchronising, which is probably part of a measure I implemented in this version to address the huge lag issue when mass applying EPGP modifications. I'll look into this as soon as I can to have it rectified. My guild is running a raid tonight so I bloody hope I can get it fixed by then. I'll keep you posted on this and your other ticket.

Thanks

**Comment by mykkenny** [comment]
I'll be honest, never noticed any lag in the previous version! But I do see tickets from others, and my PC is a powerful one so perhaps that is why.

Anything I can do to help let me know, anything I can do to prevent this happening tomorrow, let me know! Thanks for the quick response, great work with this :)

**Comment by Alumian** [comment]
Thanks man, bit embarrassing that this happened though lol. The lag was caused when you did like award guild EP and the guild had more than 200 players in it, the game engine would spam the event GUILD_ROSTER_UPDATE and the addon would try its best to catch up. But that meant it tried to essentially run 200+ cycles, so I applied a time gate to the function but now it looks like that has caused my various rosters to desynchronise haha.

Thanks again, I appreciate the offer but I'm assuming that this kind of issue is only going to show up in large scale testing so it looks like I'll have to subject my poor raid team to this to work out the solution ._.

**Comment by Alumian** [comment]
Hey Mykkenny,
I've applied a fix for this and tested it in Onyxia's Lair and Molten Core tonight, seems to be working. I'll try to have this update done ASAP but I've still got one other bug I still need to resolve your other bug.

**Comment by mykkenny** [comment]
Awesome, going MC tonight in about 7hrs so if the change is on GitHub I can pull it down and test for you if you haven't updated officially by the on curse. Nice work :)

**Comment by Alumian** [comment]
New version is up.

**Comment by PibbzTV** [comment]
I believe this fix was the cause of havoc for us last night. Everyones EP and GP was shuffled around. Glad i saved a backup before the raid to put it that way :)

We had two raids running simultaneously, and suddenly everything was wrong. My guess is that at one point we probably killed a boss at the same time, causing both loot masters to try to award auto EP for the kill before the first job had finished cycling.

Not sure tho, just a theory. Either way, my guild is big, like 700 members, so it might be part of the problem.

**Comment by PibbzTV** [comment]
Can I ask if this fix could have any impact for my situation where we run two simultaneous raids?
Would it be possible for one raid to have initiated an EP update, and it not being finished before the second raid triggers a similar update? Could that cause problems?

**Comment by mykkenny** [comment]
This seems to have fixed the issue for me. Did a raid last night with the updated version, no problems with the wrong players getting EP or GP as far as I could tell.

**Comment by Alumian** [comment]
@PibbzTV it's possible, but the only case where that would happen is if two encounters were defeated more or less at the exact same time. The roster built into CEPGP should be receiving the prompt when any changes are made to members in the guild and it'll run a full update cycle in <0.05 seconds. So unless you're killing 2 bosses within <0.05 seconds of each other lol, it's unlikely that simultaneous raids would cause this.
No, this problem was caused by a time gate I applied to the roster updates which essentially stalled the update process. In this update, I removed the time gate and implemented a different measure to ensure the clients don't lock up when a mass function is applied.
